uaclient/entitlements/base.py

```diff
@@ -358,6 +423,51 @@ def handle_incompatible_services(self) -> bool:
 
         return True
 
+    def handle_required_services(self) -> bool:
+        """
+        Prompt user when required services are found during enable.
+
+        When enabling a service, we may find that there are required services
+        that must be enabled first. In that situation, we can ask the user
+        if the required service should be enabled before proceeding.
+        """
+        from uaclient.entitlements import ENTITLEMENT_CLASS_BY_NAME
+
+        for required_service in self.required_services:
+            ent_cls = ENTITLEMENT_CLASS_BY_NAME.get(required_service)
+
+            if ent_cls:
+                ent = ent_cls(self.cfg, allow_beta=True)
```

Comment:
We're doing a lot of instantiation of ENTITLEMENT_CLASS_BY_NAME across the check_required_services (called by can_enable) vs handle_required_services vs _disable_dependent_services. I'm wondering if it's worth us persisting the instantiated classes for reference somewhere to avoid multiple calls to ent_cls(self.cfg, allow_beta=True) etc. Or whether it's worth us reading this from status cache, so we don't have to re-inspect application_status multiple times on a single enable or disable operation. In either case, it does seem like there is a lot of boilerplate code that could be potentially shared between handle_required_services, _disable_dependent_services and handle_incompatible_services. Maybe it's just as simple as a get_service_status function that could leverage a cached _service_status somehow so that all callers could just inspect that app status value and react appropriately instead of  performing the same ent_cls().application_status setup in multiple places.

Reply:
I thought about refactoring those methods, the problem is that we rely on the services through the method. We use to define the messages we are going to show to the user and we use it to perform the enable/disable operation. We could have an auxiliary function that return the entity, which will remove the import of the ENTITLEMENT_CLASS_BY_NAME dict from those methods, but I don't think it is the best refactoring we can do.

Regarding the use of status cache, we could leverage it. The problem is that we will always need to loop on the service list to find our service status, which is not ideal; Besides that, we will still need to instantiate the service anyway for the disable/enable operations, so I don't think that change will improve the code.

But I do agree that currently, those methods are sharing a lot of code and if you have any other refactoring idea or want to further discuss the proposed ones, let me know

Reply:
Yeah let's postpone this refactor idea until we can better spec out the implications and improvements to be had as caching would involve also potential cache-miss logic/errors etc.